README.md
# Python client for PhotoPrism 
A Python client to interact with photoprism.

Changelog can be found in [here](./changelog.md).

## IMPORTANT
This Python client is build on the undocumented API of PhotoPrism. There may be bugs in there. Please let me know if you find some. 

Currently tested on the [version of the 17th of June 2022](https://github.com/photoprism/photoprism/releases/tag/220617-0402b8d3). It will probably work with later versions. Just try it and let me know, I will update the readme accordingly.  

## Requirements
- requests (latest)

## Setup
To start interacting with Photoprism set up a session. You always need to do this if you start the script. Otherwise there is no session for the client to interact with PhotoPrism.
``` python
from photoprism.Session import Session
pp_session = Session("admin", "changethis", "demo.photoprism.app")
pp_session.create()
```

### HTTPS
It is possible to connect to Photoprism using HTTPS. Two variables exist for this:
- use_https: Boolean
- verify_cert: Boolean

By default, a connection will be made using HTTP. A session can be set up using HTTPS by setting the use_https variable to _True_. By default the certificate will be verified. This check can be disabled by setting the verify_cert variable to _False_. Only change this if you understand what this means. It is better to leave this variable unchanged. 

An example is:
```python
from photoprism.Session import Session
pp_session = Session("admin", "changethis", "demo.photoprism.app", use_https=True, verify_cert=False)
pp_session.create()
```

### User agent
The user agent is _Photoprism Python Client_ by default. You can change this using the variable _user_agent_. 
For example:
```python
from photoprism.Session import Session
pp_session = Session("admin", "changethis", "demo.photoprism.app", user_agent="Hello World! This is an example.")
pp_session.create()
```

## Searching
To search for photos. With this example it will return the first 100 results. You can change this with `count=1000`.

```python
from photoprism.Photo import Photo

p = Photo(pp_session)
p.search(query="original:*")
```

## Other functions
This is a list of all other functions within the client. If you want other functions either request them or send a pull request.
| function | description | variables | returns |
| ---- | ----- | ------ | ----- |
| Photo.add_photos_to_album() | Add photos to an album, you will need to provide a list of UIDs of the photos you want to add. Returns True if successfull | photos: list, album_uid: string | True if successfull  |
| Photo.add_to_album_from_query() | Provide a search query and add all photos that are returned into an album. Provide the albumname, not the UID of the album. | query: string, albumname: string, count: int (default=1000000), offset: int (default=0), order: string (default="newest") | True if successfull |
| Photo.check_if_album_exists() | Small function to check if an album exists | name: string, create_if_not: bool (default is False) | True if it exists, False if not (will continue to be False if the album is created) |
| Photo.create_album() | Create an album, returns a boolean if it worked | title: string | Dict object with the album information |
| Photo.download_album() | Download an entire album as ZIP. The file will get the same name as the title of the album, by setting the filename variable the name can be changed. | uid: string, path: string (default="."), filename: string (default=None) | True if successfull |
| Photo.download_file() | Download a single file. The file will get the same name as in photoprism, by setting the filename variable the name can be changed. | hash: string, path: string (default="."), filename: string (default=None) | True if succesfull |
| Photo.download_files_from_query() | Download files from a query | query: string, count: int (default=100), offset: int (default=0), order: string (default="newest") | True if succesfull | 
| Photo.get_album() | Get all information of an album based upon the UID of the album | uid: string | Dict object with the information of the album, False if it does not exist |
| Photo.get_album_uid_by_name() | Get the UID of an album using the name of the album. Be aware, it uses the list_albums function that is limited to 100000 albums | name: string | String of uid, None if it does not exist |
| Photo.get_uid_list_of_search() | Return a list of UIDs based upon the search | query: string, count: int (default is 100), offset: int (default=0), order: string (default="newest") | list of uids |
| Photo.list_albums() | Provide a list of all albums within the photoprism instance, with a max of 100000 | None | Dict object with all albums and their metadata (max of 100000 results) |
| Photo.raw_call() | Function to perform a request to the photoprism server (usually not needed by a user) | endpoint: string, type: string (default="GET"), data: string (default=False) | requests object |
| Photo.search() | Create the session | query: string, count: int (default=100), offset: int (default=0), order: string (default="newest") | Dict object of the results of the search |
| Photo.start_import() | Start an import job, default path is upload. It returns True when the import started, not when finished | path: string (default="upload"), move: Bool (default=False) | True if successfully started |
| Photo.stop_import() | Stop an import job | None | True if successfully stopped |
| Photo.remove_photos_from_album() | Remove photos from an album, Returns True if successfull | albumname: string, photos: bool (default=False), count: int (default=1000000) | True if succesfull |
| Photo.remove_album() | Remove album based on album name | albumname: string | Dict data returned from the server |
| Photo.remove_album_uid() | Remove album based on album uid | uid: string | Dict of the removed album, False if action failed. |


## License 
MIT License

Copyright (c) 2022 maartenvl

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


changelog.md
# Changelog
Latest update: May 9th 2023
## Version 0.2
### Additions
- Added a count for Photo.add_to_album_from_query() for a maximum count that are from the query, default is 1000000.
- The ability to download files and albums has been added:
	- Photo.download_files(): Download a single file.
	- Photo.download_album(): Download an entire album as ZIP.
	- Photo.download_files_from_query(): Download files from a query.
- Added the ability to force the use of HTTPS. As per request of: [cluk33](https://github.com/cluk33) ([issue #3](https://github.com/mvlnetdev/photoprism_client/issues/3))
- Added offset en order variables to Photo.search() ([PR #4](https://github.com/mvlnetdev/photoprism_client/pull/4) and [PR #5](https://github.com/mvlnetdev/photoprism_client/pull/5)). Thanks [frascu](https://github.com/frascu)!
- Added two functions to remove a photo from an album and to remove an album ([PR #2](https://github.com/mvlnetdev/photoprism_client/pull/2)).  Thanks [sajben94](https://github.com/sajben94)!
- Added the ability to change the User Agent.

### Miscellaneous
- Photo.create_album() now returns a dict object of the newly created album.
- examples/manage_albums.py now contains all functions for albums.
- examples/download_files.py is added to show examples to download files.



photoprism/Session.py
"""Set up a session with the photoprism server"""

import json
from requests import Session as r_session, Request as r_request

from photoprism import mimetypes

class Session():
    def __init__(self, username, password, host, use_https=False, verify_cert=True, user_agent=None):
        """Initialize using a username, password and hostname"""
        self.username = username
        self.password = password
        self.login_data = {
            "username": self.username,
            "password": self.password
        }

        self.host = host

        self.protocol = "http"
        self.ssl_enable = use_https
        if self.ssl_enable:
            self.protocol = self.protocol + "s"
            self.ssl_verify = verify_cert
        self.url = f"{self.protocol}://{self.host}/api/v1"


        self.user_agent = "Photoprism Python Client"
        if user_agent:
            self.user_agent = user_agent

        self.headers = {
            "User-Agent": self.user_agent
        }

    def req(self, endpoint, method, stream=False, **kwargs):
        s = r_session()
        r = r_request(method=method, url=f"{self.url}{endpoint}")

        if self.ssl_enable:
            s.verify = self.ssl_verify
        r.headers = self.headers

        for k, v in kwargs.items():
            if k == "data":
                r.data = json.dumps(v)

        p = r.prepare()
        resp = s.send(p, stream=stream)

        headers = resp.headers
        content_type = headers["Content-Type"].split("; ")
        mime_type, mime_subtype = content_type[0].split("/")

        # Process the return data
        if mime_type == "application":
            if mime_subtype == "json":
                data_out = json.loads(resp.text)
            elif mime_subtype == "zip":
                filename = self.determine_filename(kwargs, mime_type, mime_subtype, headers)
                with open(filename, 'wb') as f:
                    for chunk in resp.iter_content(chunk_size=10240):
                        f.write(chunk)
                data_out = True
            else:
                data_out = resp.text

        # Process downloaded single files
        elif mime_type in ["image", "video"]:
            filename = self.determine_filename(kwargs, mime_type, mime_subtype, headers)

            with open(filename, "wb") as f:
                resp.raw.decode_content = True
                f.write(resp.content)
            data_out = True

        else:
            data_out = resp.text

        return resp.status_code, data_out

    def create(self):
        """Create the session"""
        _, data = self.req("/session", "POST", data=self.login_data)
        self.session_id = data["id"]
        self.headers["X-Session-ID"] = self.session_id
        return True

    def determine_filename(self, args, mime_type, mime_subtype, headers):
        if args["filename"] != None:
            extension = mimetypes.type[mime_type][mime_subtype]
            filename = f"{args['filename']}{extension}"
        else:
            header_filename = headers["Content-Disposition"].split("; ")[1].split("=")[1]
            # Sometimes the filename in the header is enclosed, sometimes it isn't. This is to account for that.
            if header_filename[0] == '"' and header_filename[-1:] == '"':
                filename = header_filename[1:-1]
            else:
                filename=header_filename

        if "path" in args:
            filename = f"{args['path']}/{filename}"

        return filename


examples/manage_albums.py
import uuid
from photoprism.Session import Session
from photoprism.Photo import Photo

pp_session = Session("admin", "changethis", "demo.photoprism.app")
pp_session.create()

p = Photo(pp_session)

# List all albums
albums = p.list_albums()

# Check if an album exists
if p.check_if_album_exists("This album probably does not exists") == False:
    print("Does not exists")
else:
    print("It does exist")

# Show the info of an album
album_uid=albums[0]["UID"]
print(f"The name of the first album is: {p.get_album(album_uid)['Title']}")

# Create an album
temp_name=str(uuid.uuid4())
new_album = p.create_album(temp_name)
print(new_album)

# Add the first photo to the album
foto_uid = p.search(query="original:*", count=1)
print(p.add_photos_to_album([foto_uid[0]["UID"]],new_album["UID"]))

print(p.remove_photos_from_album(albumname=new_album["Title"]))

print(p.add_to_album_from_query(query="original:*", albumname=new_album["Title"], count=1))

# Delete an album
print(p.remove_album(temp_name))

photoprism/Photo.py
"""Interaction with the API of PhotoPrism"""
import uuid

from photoprism import Session
import requests, json

class Photo():
    def __init__(self, session):
        """Initialize based upon a session"""
        if type(session) != Session.Session:
            raise TypeError(f"session variable is not of type photoprism.Session.Session, but {type(session)}")

        self.session = session

    def search(self, query, count=100, offset=0, order="newest"):
        """Basic search function. Returns a Dict object based upon the returned JSON"""
        _, data = self.session.req(f"/photos?count={count}&q={query}&offset={offset}&order={order}", "GET")
        return data

    def get_uid_list_of_search(self, query, count=100, offset=0, order="newest"):
        """Return a list of UIDs based upon the search"""

        photos = self.search(query, count, offset=offset, order=order)
        photolist = []
        for ps in photos:
            photolist.append(ps["UID"])

        return photolist

    def list_albums(self, count=100000):
        """Provide a list of all albums within the photoprism instance, with a max of 100000"""
        _, data = self.session.req(f"/albums?count={count}", "GET")
        return data

    def check_if_album_exists(self, name, create_if_not=False):
        """Small function to check if an album exists"""

        data = self.list_albums()
        for d in data:
            if name == d["Title"]:
                return True

        if create_if_not:
            self.create_album(name)

        return False

    def get_album_uid_by_name(self, name):
        """Get the UID of an album using the name of the album. Be aware, it uses the list_albums function that is limited to 100000 albums"""

        data = self.list_albums()
        uid = None
        for d in data:
            if name == d["Title"]:
                return d["UID"]
        
        return False

    def create_album(self, title):
        """Create an album, returns a boolean if it worked"""
        data = {"Title":title,"Favorite":False}
        status_code, output = self.session.req("/albums", "POST", data=data)

        if status_code == 200:
            return output
        
        return False

    def add_photos_to_album(self, photos, album_uid):
        """Add photos to an album, you will need to provide a list of UIDs of the photos you want to add. Returns True if successfull"""
        data = {
            "photos":photos
        }
        status_code, _ = self.session.req(f"/albums/{album_uid}/photos", "POST", data=data)
        if status_code == 200:
            return True

        return False

    def add_to_album_from_query(self, query, albumname, count=1000000, offset=0, order="newest"):
        """Provide a search query and add all photos that are returned into an album. Provide the albumname, not the UID of the album."""

        self.check_if_album_exists(albumname, create_if_not=True)
        album_uid = self.get_album_uid_by_name(albumname)
        photolist = self.get_uid_list_of_search(query, count=count, offset=offset, order=order)
        result = self.add_photos_to_album(photolist, album_uid)
        return result

    def get_album(self, uid):
        """Get all information of an album based upon the UID of the album"""
        status_code, data = self.session.req(f"/albums/{uid}", "GET")
        if status_code == 200:
            return data

        return False

    def remove_photos_from_album(self, albumname, photos=False, count=1000000):
        """Remove photos from an album, Returns True if successfull"""
        album_uid = self.get_album_uid_by_name(albumname)
        if photos == False:
            query = f"album:\"{albumname}\""
            photos = self.get_uid_list_of_search(query,count=count)

        data = {
            "photos":photos
        }
        status_code, _ = self.session.req(f"/albums/{album_uid}/photos", "DELETE", data=data)
        if status_code == 200:
            return True

        return False

    def remove_album(self, albumname):
        """Remove album based on album name"""
        album_uid = self.get_album_uid_by_name(albumname)
        return self.remove_album_uid(album_uid)

    def remove_album_uid(self, uid):
        """Remove album based on album uid"""
        status_code, data = self.session.req(f"/albums/{uid}", "DELETE")
        if status_code == 200:
            return data

        return False

    def start_import(self, path="upload", move=False):
        """Start an import job, default path is upload. It returns True when the import started, not when finished"""
        data = {
            "path": path,
            "move": move
        }
        status_code, _ = self.session.req(f"/import", "POST", data=data)
        if status_code == 200:
            return True
        
        return False

    def stop_import(self):
        """Stop an import job"""
        status_code, _ = self.session.req("/import", "DELETE")
        if status_code == 200:
            return True
        
        return False

    def download_file(self, hash, path=".", filename=None):
        """Download a single file"""
        status_code, data = self.session.req(f"/dl/{hash}", "GET", path=path, filename=filename)

        if status_code == 200:
            return True
        return False

    def download_album(self, uid, path=".", filename=None):
        """Download an entire album as ZIP"""
        status_code, data = self.session.req(f"/albums/{uid}/dl", "GET", stream=True, path=path, filename=filename)

        if status_code == 200:
            return True
        return False

    def download_files_from_query(self, query, count=100, offset=0, order="newest"):
        """Download files from a query"""
        # In order to make it easy, files are first put into a temporary album
        # After which they are downloaded
        # After the download is complete, the folder is removed.

        temp_album = self.create_album(str(uuid.uuid4()))
        status = self.add_to_album_from_query(query, temp_album["Title"], count=count, offset=offset, order=order)
        status = self.download_album(temp_album["UID"])
        status = self.remove_album_uid(temp_album["UID"])
        return status




